File: cpp/lc46.cxx
#include <vector>
#include <set>
using namespace std;
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> paths;
        vector<int> path;
        set<int> used;
        traverse(path, paths, used, nums);
        return paths;
    }

    void traverse(vector<int>& path, vector<vector<int>>& paths, set<int>& used, vector<int>& nums) {
        if (nums.size() == path.size()) {
            paths.push_back(vector<int>(path));
            return;
        }

        for (auto n: nums) {
            if (used.find(n) != used.end()) continue;
            used.insert(n);
            path.push_back(n);
            traverse(path, paths, used, nums);
            path.pop_back();
            used.erase(n);
        }
    }
};

File: src/bin/lc46.rs
use std::collections::HashSet;

struct Solution;
fn main() {}
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut paths = vec![];
        let mut path = vec![];
        let mut choosen = HashSet::new();
        Self::traverse(&mut path, &mut paths, &nums, &mut choosen);
        paths
    }

    fn traverse(path: &mut Vec<i32>, paths: &mut Vec<Vec<i32>>, nums: &Vec<i32>, choosen: &mut HashSet<i32>) {
        if nums.len() == choosen.len() {
            paths.push(path.clone());
            return;
        }

        for num in nums {
            if choosen.contains(num) {
                continue;
            }
            path.push(*num);
            choosen.insert(*num);
            Self::traverse(path, paths, nums, choosen);
            choosen.remove(num);
            path.pop();
        }
    }
}